Superfluous rest created when when Music21 interprets a bad MusicXML

<!-- Version of music21:  print(music21.__version__) -->
**music21 version**

`6.1 through 6.5, most likely other versions `   <!-- replace with actual version -->


**Problem summary**
<!-- Briefly: what undesired thing happens when what action is taken? -->

**TLDR:** 
When a MusicXML measure contains elements with more duration than should be specified by the time signature, rests are incorrectly created and appended to the beginning of the measures, with duration equal to the aforementioned duration discrepancy. Because of this, Music21 content no longer becomes equal to MusicXML content. More issues arise when the durations are shorter, but they are less important issues.

**Steps to reproduce**

I'm adding external files because the issue stems from bad XML files themselves, not anything that can easily be reproduced by creating a stream manually.
[Rachmaninoff_Prelude_Op._23_No._5_G_Minor.mxl.zip](https://github.com/cuthbertLab/music21/files/5821344/Rachmaninoff_Prelude_Op._23_No._5_G_Minor.mxl.zip)



```
import music21
m21_stream = music21.converter.parse('/path/to/MusicXML/Rachmaninoff_Prelude_Op._23_No._5_G_Minor.mxl')
m21_stream[3][190].show('t') ###Measure 6, Part P1-Staff1
m21_stream[4][190].show('t') ###Measure 6, Part P1-Staff2
### To view MusicXML content, my suggestion is to use MuseScore or Finale. OSMD was okay, but didn't seem as clear or accurate as MuseScore or Finale.
```

**Expected vs. actual behavior**
**Expected**:
When Music21 reads in a bad MusicXML where elements have duration longer than the duration specified by the XML time signature, Music21 converts these measures and deals with the MusicXML errors however it chooses to, but does not add any incorrect or superfluous content to the stream measures when this occurs.

**Actual**:
When Music21 reads in a bad MusicXML where elements have duration longer than the duration specified by the XML time signature, Music21 converts these measures and creates the elements as written with overextended duration, but also creates extra incorrect rest elements that were not in the XML, appended to the beginning of measures and voices.

**More information**

**Long read, and my thought process:**
This shouldn't have to be a problem, but it's one I've unfortunately run into.

In a MusicXML, when the duration of a part's measure's elements extended beyond the expected duration of the measure, a funny phenomenon occurs in the XML to M21 conversion. If an instrument has multiple staves or voices, for every staff/voice that is "next" after an overextended measure, a rest is added to the beginning of the next staff/voice *equal to the duration by which the last staff/voice had been overextended. Here's a concrete example:

In an included piece by Rachmaninoff, measure 6, there are two staves and multiple voices per part. 
**In the MusicXML:**
- In measure 6, staff 1, voice 1, there are 4 beats worth of elements, which is fine.
- In staff 1, voice 2, there are 4.5 beats worth of elements, which will later cause a cascade of issues.
- Voices 5 and 6 are not present in staff 1, and are written after the duration error in the MusicXML.
- In measure 6, staff 2, voice 5, there are 4 beats worth of elements, but a duration issue in this part did precede.
- In staff 1, voice 6, there are 4.5 beats worth of elements, on top of the preceding duration error.
- Voices 1 and 2 are not present in staff 2, and are written before or during the duration error in the MusicXML.

**In the M21 stream:**
- In measure 6, staff 1, voice 1, the elements in this staff have the right duration, so we're okay. 
- In staff 1, voice 2, the note D has a duration of 1.5, and the rest has a duration of 3, making 4.5 beats total, 0.5 beats over. The handling of this measure I believe is okay. However, all future voices/staves will start with a rest of 0.5 duration.
- In staff 1, voice 5, this voice was previously absent in the staff, but because it exists in the MusicXML part, and I believe Music21 added it to this staff for the purpose of adding the extra 0.5 duration rest.
- Staff 1, voice 6 is the same as above. Note that neither voice 5 nor 6 extend beyond the time signature duration, so they don't affect future staves/voices.
- In measure 6, staff 2, voice 1, because MusicXML's denote two staves of a piano as the *same* part, Music21 will have the same issue of offsetting staff 2 elements by 0.5 because of what happened in staff 1, but now they're just separated into different staves. Staff 2, voice 5 now has an extra rest in the front. Because the MusicXML itself had 4 beats of elements, the offset of the next voice is not affected beyond the same 0.5-beat offset and 0.5-duration rest added in the beginning.
- In measure 6, staff 2, voice 6, the same 0.5-beat offset occurred, resulting in the same 0.5-duration rest added in the beginning.
- Voices 1 and 2 don't show up in the second staff because by the time the duration error occurred, they were passed by in the MusicXML.
- Let's say, you had a 7th voice after voice 6. After testing, I have seen that, in staff 2 voice 7, there would be a 1.0-beat rest added to compensate for the 0.5-beat duration error in staff 1 voice 2 and the 0.5-beat duration error in staff 2 voice 6 accumulating.

```
###Time signature 4/4
(Pdb) m21_stream[3][190].show('t') ### Measure 6, Part P1-Staff1
{0.0} <music21.stream.Voice 1>
    {0.0} <music21.note.Rest rest>
    {0.5} <music21.chord.Chord D4 F#4 D5>
    {0.75} <music21.chord.Chord D4 F#4 D5>
    {1.0} <music21.chord.Chord D4 F#4 D5>
    {1.5} <music21.note.Note F#>
    {1.75} <music21.note.Note A>
    {2.0} <music21.note.Note B->
    {2.5} <music21.chord.Chord G3 D4 E4>
    {3.0} <music21.note.Note G>
    {3.5} <music21.chord.Chord B-3 D4 E4>
{0.0} <music21.stream.Voice 2>
    {0.0} <music21.note.Note D>
    {1.5} <music21.note.Rest rest> ###duration of 3 beats, total voice duration of 4.5 beats
{0.0} <music21.stream.Voice 5>
    {0.0} <music21.note.Rest rest> ### duration of 0.5 = 4.5 - 4.0 beats
{0.0} <music21.stream.Voice 6>
    {0.0} <music21.note.Rest rest> ### duration of 0.5 = 4.5 - 4.0 beats
{1.5} <music21.clef.BassClef>
{2.5} <music21.clef.TrebleClef>

(Pdb) m21_stream[4][190].show('t') ### Measure 6, Part P1-Staff2
{0.0} <music21.stream.Voice 5>
    {0.0} <music21.note.Rest rest> ### duration of 0.5 = 4.5 - 4.0 beats
    {0.5} <music21.note.Rest rest>
    {1.0} <music21.chord.Chord A2 F#3 A3>
    {1.25} <music21.chord.Chord A2 F#3 A3>
    {1.5} <music21.chord.Chord A2 F#3 A3>
    {2.0} <music21.note.Note F#>
    {2.25} <music21.note.Note A>
    {2.5} <music21.note.Note B->
    {3.0} <music21.chord.Chord D3 E3>
    {3.5} <music21.chord.Chord G1 G2>
    {4.0} <music21.chord.Chord D3 E3> ###duration of 0.5 beats, total voice duration of 4.5 beats
{0.0} <music21.stream.Voice 6>
    {0.0} <music21.note.Rest rest> ### duration of 0.5 = 4.5 - 4.0 beats
    {0.5} <music21.chord.Chord D1 D2>
    {2.0} <music21.note.Rest rest>  ###duration of 3 beats, total voice duration of 5 beats
```

I didn't finish diagnosing the problem, but it occurs in music21/musicxml/xmltom21.py, around line 2320-2340 in measureParser.parse(), probably at the v.makeRests() function. Before this, the stream contains elements only from the MusicXML (as expected), but all elements (at least GeneralNotes) were offset by the amount of beats the last staff/voice went over by, e.g. a voice object looking like this.

```
###Time signature 4/4
{0.0} <music21.stream.Voice 5>
    {0.5} <music21.note.Rest rest>
    {1.0} <music21.chord.Chord A2 F#3 A3>
    {1.25} <music21.chord.Chord A2 F#3 A3>
    {1.5} <music21.chord.Chord A2 F#3 A3>
    {2.0} <music21.note.Note F#>
    {2.25} <music21.note.Note A>
    {2.5} <music21.note.Note B->
    {3.0} <music21.chord.Chord D3 E3>
    {3.5} <music21.chord.Chord G1 G2>
    {4.0} <music21.chord.Chord D3 E3>
```
My guess is that the coreInsert() uses some kind of info written in the XML to decide how far to backtrack the offset when creating streams, and in this case, it didn't back up far enough because the last measure's elements went over the expected duration. Then 4.5 current offset of the last voice's 

After the v.makeRests() call, the voice would look like this:

```
###Time signature 4/4
{0.0} <music21.stream.Voice 5>
    {0.0} <music21.note.Rest rest>
    {0.5} <music21.note.Rest rest>
    {1.0} <music21.chord.Chord A2 F#3 A3>
    {1.25} <music21.chord.Chord A2 F#3 A3>
    {1.5} <music21.chord.Chord A2 F#3 A3>
    {2.0} <music21.note.Note F#>
    {2.25} <music21.note.Note A>
    {2.5} <music21.note.Note B->
    {3.0} <music21.chord.Chord D3 E3>
    {3.5} <music21.chord.Chord G1 G2>
    {4.0} <music21.chord.Chord D3 E3>
```
and suddenly the stream has rest elements that don't match the content in the MusicXML.

I was able to write a work-around routine to deal with this. I use a regular expression to strip the staff label from the part ID's, so part ID's P1-Staff1 and P1-Staff2 both become P1, P1 and match each other. After this, I go sequentially through the measures/voices until I find at least one duration overshoot. After this, I delete the first element of every staff/voice remaining to that part, because I assume the element will be a superfluous rest.

Although this routine was easy enough to write, I write this report so that other people are aware of the issue and so maybe it can be rectified in the source code. I'd also love to explore this more and see if I can fix in source, if you happen to be busy. I've already spent a couple days with this problem.

There's a chance this might actually run faster. I'm showing about a 1-5% improvement (depending on the tools used, I'm not sitting here for an hour to run 1m loops) if we just use the raw duration, and then get a DurationTuple from the cooked version and short circuit immediately if equal.

The overhead is probably saved from not having to clear and then readd the duration tuple in the cooked workflow.

```diff
diff --git a/music21/musicxml/xmlToM21.py b/music21/musicxml/xmlToM21.py
index 781f39705..4fe603615 100644
--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3234,8 +3234,9 @@ class MeasureParser(XMLParserBase):
             forceRaw = True
             # TODO: empty-placement
 
+        rawQL = qLen
         # two ways to create durations, raw (from qLen) and cooked (from type, time-mod, dots)
-        if forceRaw:
+        if True:
             if d is not None:
                 # environLocal.printDebug(['forced to use raw duration', durRaw])
                 durRaw = duration.Duration()  # raw just uses qLen
@@ -3258,11 +3259,15 @@ class MeasureParser(XMLParserBase):
                     #                         'mxNote:',
                     #                         ET.tostring(mxNote, encoding='unicode'),
                     #                         'last mxDivisions:', divisions])
-                    durRaw.quarterLength = qLenRounded
+                    rawQL = durRaw.quarterLength = qLenRounded
             else:
                 d = duration.Duration(quarterLength=qLen)
-        else:  # a cooked version builds up from pieces
+        # can't do this unless we have a type, so if not forceRaw
+        if not forceRaw:  # a cooked version builds up from pieces
             dt = duration.durationTupleFromTypeDots(durationType, numDots)
+            if dt.quarterLength == rawQL:
+                # equality passed, so we're done
+                return d if inputM21 is None else None
             if d is not None:
                 d.clear()
                 d.addDurationTuple(dt)

```

timing script (can't use corpus.parse, because that doesn't accept writePickle=False):

```
from music21 import *
cmd = "converter.parse('music21/corpus/beach/prayer_of_a_tired_child.musicxml', forceSource=True, storePickle=False)"
import cProfile
cProfile.run(cmd)
```


Something breaks with tuplets if I do this, so, of course, all bets are off timing-wise until we have a real patch. Not going to jump on this, but found it curious.

```
======================================================================
FAIL: testNestedTuplets (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jwalls/music21/music21/musicxml/xmlToM21.py", line 6571, in testNestedTuplets
    self.assertEqual(repr(nList[i].duration.tuplets),
AssertionError: '(<music21.duration.Tuplet 15/8/256th>,)' != '(<music21.duration.Tuplet 3/2/eighth>, <music21.duration.Tuplet 5/2/eighth>)'
- (<music21.duration.Tuplet 15/8/256th>,)
+ (<music21.duration.Tuplet 3/2/eighth>, <music21.duration.Tuplet 5/2/eighth>)
```

--- a/music21/_version.py
+++ b/music21/_version.py
@@ -42,7 +42,7 @@ When changing, update the single test case in base.py.
 Changing this number invalidates old pickles -- do it if the old pickles create a problem.
 '''
 
-__version_info__ = (7, 0, 3)  # can be 4-tuple: (7, 0, 5, 'a2')
+__version_info__ = (7, 0, 4)  # can be 4-tuple: (7, 0, 5, 'a2')
 
 v = '.'.join(str(x) for x in __version_info__[0:3])
 if len(__version_info__) > 3 and __version_info__[3]:

--- a/music21/base.py
+++ b/music21/base.py
@@ -28,7 +28,7 @@ available after importing `music21`.
 <class 'music21.base.Music21Object'>
 
 >>> music21.VERSION_STR
-'7.0.3'
+'7.0.4'
 
 Alternatively, after doing a complete import, these classes are available
 under the module "base":

--- a/music21/musicxml/testTab.xml
+++ b/music21/musicxml/testTab.xml
@@ -9,6 +9,12 @@
     </part-list>
     <part id="P1">
         <measure number="0">
+            <attributes>
+                <time>
+                    <beats>4</beats>
+                    <beat-type>4</beat-type>
+                </time>
+            </attributes>
             <note>
                 <type>quarter</type>
                 <notations>

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3199,8 +3199,7 @@ class MeasureParser(XMLParserBase):
 
         If the `<duration>` doesn't match the `<type>` and `<dots>`,
         an unlinked duration is created so that `.quarterLength` agrees with
-        `<duration>` but the notated types can still be represented by
-        `.components`.
+        `<duration>` but the notated types can still be represented.
 
         Create a second dot on `mxNote` and parse again, observing the identical
         `quarterLength`:

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -6515,7 +6515,7 @@ class Test(unittest.TestCase):
           <step>D</step>
           <octave>5</octave>
         </pitch>
-        <duration>{defaults.divisionsPerQuarter * 0.5}</duration>
+        <duration>{defaults.divisionsPerQuarter * 0.5 * (2/3) * (2/3)}</duration>
         <voice>1</voice>
         <type>eighth</type>
         <time-modification>

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3259,8 +3259,9 @@ class MeasureParser(XMLParserBase):
         # can't do this unless we have a type, so if not forceRaw
         if not forceRaw:  # a cooked version builds up from pieces
             dt = duration.durationTupleFromTypeDots(durationType, numDots)
-            if dt.quarterLength == qLen:
+            if dt.quarterLength == qLen and not tuplets:
                 # raw == cooked, so we're done
+                # but this comparison gives false positives if tuplets are involved
                 return d if inputM21 is None else None
             if d is not None:
                 d.clear()
@@ -3271,6 +3272,7 @@ class MeasureParser(XMLParserBase):
 
             for tup in tuplets:
                 d.appendTuplet(tup)
+                qLen = common.opFrac(qLen * tup.tupletMultiplier())
 
             # Second check against qLen (raw), now with tuplets
             # if != , create unlinked Duration and set raw qLen
@@ -6506,13 +6508,13 @@ class Test(unittest.TestCase):
         '''
         test that a note with nested tuplets gets converted properly.
         '''
-        mxN = '''
+        mxN = f'''
         <note default-x="347">
         <pitch>
           <step>D</step>
           <octave>5</octave>
         </pitch>
-        <duration>4</duration>
+        <duration>{defaults.divisionsPerQuarter * 0.5}</duration>
         <voice>1</voice>
         <type>eighth</type>
         <time-modification>

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3235,42 +3235,48 @@ class MeasureParser(XMLParserBase):
             # TODO: empty-placement
 
         # two ways to create durations, raw (from qLen) and cooked (from type, time-mod, dots)
-        if forceRaw:
-            if d is not None:
-                # environLocal.printDebug(['forced to use raw duration', durRaw])
-                durRaw = duration.Duration()  # raw just uses qLen
-                # the qLen set here may not be computable, but is not immediately
-                # computed until setting components
-                durRaw.quarterLength = qLen
-                try:
-                    d.components = durRaw.components
-                except duration.DurationException:  # TODO: Test
-                    if qLen:
-                        qLenRounded = 2.0 ** round(math.log2(qLen))
-                    else:
-                        qLenRounded = 0.0
-                    environLocal.printDebug(
-                        ['mxToDuration',
-                         f'rounding duration to {qLenRounded} as type is not'
-                         + 'defined and raw quarterLength '
-                         + f'({qLen}) is not a computable duration'])
-                    # environLocal.printDebug(['mxToDuration', 'raw qLen', qLen, durationType,
-                    #                         'mxNote:',
-                    #                         ET.tostring(mxNote, encoding='unicode'),
-                    #                         'last mxDivisions:', divisions])
-                    durRaw.quarterLength = qLenRounded
-            else:
-                d = duration.Duration(quarterLength=qLen)
-        else:  # a cooked version builds up from pieces
+        if d is not None:
+            # environLocal.printDebug(['forced to use raw duration', durRaw])
+            durRaw = duration.Duration()  # raw just uses qLen
+            # the qLen set here may not be computable, but is not immediately
+            # computed until setting components
+            durRaw.quarterLength = qLen
+            try:
+                d.components = durRaw.components
+            except duration.DurationException:  # TODO: Test
+                if qLen:
+                    qLenRounded = 2.0 ** round(math.log2(qLen))
+                else:
+                    qLenRounded = 0.0
+                environLocal.printDebug(
+                    ['mxToDuration',
+                        f'rounding duration to {qLenRounded} as type is not'
+                        + 'defined and raw quarterLength '
+                        + f'({qLen}) is not a computable duration'])
+                durRaw.quarterLength = qLen = qLenRounded
+        else:
+            d = duration.Duration(quarterLength=qLen)
+        # can't do this unless we have a type, so if not forceRaw
+        if not forceRaw:  # a cooked version builds up from pieces
             dt = duration.durationTupleFromTypeDots(durationType, numDots)
+            if dt.quarterLength == qLen:
+                # raw == cooked, so we're done
+                return d if inputM21 is None else None
             if d is not None:
                 d.clear()
+                d.tuplets = []
                 d.addDurationTuple(dt)
             else:
                 d = duration.Duration(durationTuple=dt)
 
             for tup in tuplets:
                 d.appendTuplet(tup)
+
+            # Second check against qLen (raw), now with tuplets
+            # if != , create unlinked Duration and set raw qLen
+            if d.quarterLength != qLen:
+                d.linked = False
+                d.quarterLength = qLen
 
         if inputM21 is None:
             return d

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3257,10 +3257,9 @@ class MeasureParser(XMLParserBase):
 
         # two ways to create durations, raw (from qLen) and cooked (from type, time-mod, dots)
         if d is not None:
-            durRaw = duration.Duration()  # raw just uses qLen
+            durRaw = duration.Duration(quarterLength=qLen)  # raw just uses qLen
             # the qLen set here may not be computable, but is not immediately
             # computed until setting components
-            durRaw.quarterLength = qLen
             try:
                 d.components = durRaw.components
             except duration.DurationException:  # TODO: Test

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3196,6 +3196,26 @@ class MeasureParser(XMLParserBase):
         'eighth'
         >>> c.dots
         1
+
+        If the `<duration>` doesn't match the `<type>` and `<dots>`,
+        an unlinked duration is created so that `.quarterLength` agrees with
+        `<duration>` but the notated types can still be represented by
+        `.components`.
+
+        Create a second dot on `mxNote` and parse again, observing the identical
+        `quarterLength`:
+
+        >>> from xml.etree.ElementTree import SubElement
+        >>> unused = SubElement(mxNote, 'dot')
+        >>> c2 = MP.xmlToDuration(mxNote)
+        >>> c2
+        <music21.duration.Duration unlinked type:eighth quarterLength:0.75>
+        >>> c2.quarterLength
+        0.75
+        >>> c2.type
+        'eighth'
+        >>> c2.dots
+        2
         '''
         numDots = 0
         tuplets = ()
@@ -3238,7 +3258,6 @@ class MeasureParser(XMLParserBase):
 
         # two ways to create durations, raw (from qLen) and cooked (from type, time-mod, dots)
         if d is not None:
-            # environLocal.printDebug(['forced to use raw duration', durRaw])
             durRaw = duration.Duration()  # raw just uses qLen
             # the qLen set here may not be computable, but is not immediately
             # computed until setting components

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3274,7 +3274,6 @@ class MeasureParser(XMLParserBase):
 
             for tup in tuplets:
                 d.appendTuplet(tup)
-                qLen = common.opFrac(qLen * tup.tupletMultiplier())
 
             # Second check against qLen (raw), now with tuplets
             # if not isclose , create unlinked Duration and set raw qLen

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3279,7 +3279,7 @@ class MeasureParser(XMLParserBase):
         # can't do this unless we have a type, so if not forceRaw
         if not forceRaw:  # a cooked version builds up from pieces
             dt = duration.durationTupleFromTypeDots(durationType, numDots)
-            if not tuplets and (dt.quarterLength == qLen):
+            if (dt.quarterLength == qLen) and not tuplets:
                 # raw == cooked, so we're done
                 # but this comparison gives false positives if tuplets are involved
                 # don't bother with isclose; merely trying to short-circuit

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -2577,6 +2577,8 @@ class MeasureParser(XMLParserBase):
         <music21.pitch.Pitch A3>
         >>> c.pitches[1]
         <music21.pitch.Pitch B3>
+        >>> c.duration
+        <music21.duration.Duration unlinked type:quarter quarterLength:0.75>
 
         >>> a = EL('<note><pitch><step>A</step><octave>3</octave></pitch>'
         ...        + qnDuration
@@ -3215,6 +3217,19 @@ class MeasureParser(XMLParserBase):
         'eighth'
         >>> c2.dots
         2
+
+        Grace note durations will be converted later to GraceDurations:
+
+        >>> mxDuration = mxNote.find('duration')
+        >>> mxNote.remove(mxDuration)
+        >>> mxGrace = SubElement(mxNote, 'grace')
+        >>> MP.xmlToDuration(mxNote, inputM21=c2)
+        >>> c2
+        <music21.duration.Duration unlinked type:eighth quarterLength:0.0>
+        >>> gn1 = note.Note(duration=c2)
+        >>> gn2 = MP.xmlGraceToGrace(mxGrace, gn1)
+        >>> gn2.duration
+        <music21.duration.GraceDuration unlinked type:zero quarterLength:0.0>
         '''
         numDots = 0
         tuplets = ()

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -3261,9 +3261,10 @@ class MeasureParser(XMLParserBase):
         # can't do this unless we have a type, so if not forceRaw
         if not forceRaw:  # a cooked version builds up from pieces
             dt = duration.durationTupleFromTypeDots(durationType, numDots)
-            if isclose(dt.quarterLength, qLen, abs_tol=1e-7) and not tuplets:
-                # raw ~= cooked, so we're done
+            if not tuplets and (dt.quarterLength == qLen):
+                # raw == cooked, so we're done
                 # but this comparison gives false positives if tuplets are involved
+                # don't bother with isclose; merely trying to short-circuit
                 return d if inputM21 is None else None
             if d is not None:
                 d.clear()
@@ -3276,7 +3277,7 @@ class MeasureParser(XMLParserBase):
                 d.appendTuplet(tup)
 
             # Second check against qLen (raw), now with tuplets
-            # if not isclose , create unlinked Duration and set raw qLen
+            # if not almost equal, create unlinked Duration and set raw qLen
             if not isclose(d.quarterLength, qLen, abs_tol=1e-7):
                 d.linked = False
                 d.quarterLength = qLen

--- a/music21/musicxml/xmlToM21.py
+++ b/music21/musicxml/xmlToM21.py
@@ -18,6 +18,8 @@ import re
 # import sys
 # import traceback
 import unittest
+
+from math import isclose
 from typing import List, Optional, Dict, Tuple, Set
 
 import xml.etree.ElementTree as ET
@@ -3259,8 +3261,8 @@ class MeasureParser(XMLParserBase):
         # can't do this unless we have a type, so if not forceRaw
         if not forceRaw:  # a cooked version builds up from pieces
             dt = duration.durationTupleFromTypeDots(durationType, numDots)
-            if dt.quarterLength == qLen and not tuplets:
-                # raw == cooked, so we're done
+            if isclose(dt.quarterLength, qLen, abs_tol=1e-7) and not tuplets:
+                # raw ~= cooked, so we're done
                 # but this comparison gives false positives if tuplets are involved
                 return d if inputM21 is None else None
             if d is not None:
@@ -3275,8 +3277,8 @@ class MeasureParser(XMLParserBase):
                 qLen = common.opFrac(qLen * tup.tupletMultiplier())
 
             # Second check against qLen (raw), now with tuplets
-            # if != , create unlinked Duration and set raw qLen
-            if d.quarterLength != qLen:
+            # if not isclose , create unlinked Duration and set raw qLen
+            if not isclose(d.quarterLength, qLen, abs_tol=1e-7):
                 d.linked = False
                 d.quarterLength = qLen
 

